src/imitation/algorithms/preference_comparisons.py

```diff
@@ -317,6 +317,144 @@ def _get_trajectories(
+                of returns as the probability that a fragment is preferred.
+                This is the discount factor used to calculate those returns.
+                Default is 1, i.e. undiscounted sums of rewards (which is what
+                the DRLHP paper uses).
+            threshold: the preference model used to compute the loss contains
+                a softmax of returns. To avoid overflows, we clip differences
+                in returns that are above this threshold. This threshold
+                is therefore in logspace. The default value of 50 means
+                that probabilities below 2e-22 are rounded up to 2e-22.
+        """
+        super().__init__()
+        self.model = model
+        self.discount_factor = discount_factor
+        self.noise_prob = noise_prob
+        self.threshold = threshold
+        self.is_ensemble, base_model = is_base_model_ensemble(self.model)
+        # if the base model is an ensemble model, then keep the base model as
+        # model to get rewards from all networks
+        if self.is_ensemble:
+            self.model = base_model
+
+    def forward(self, fragment_pairs: Sequence[TrajectoryPair]):
+        """Computes the preference probability of first fragment for all pairs.
+
+        Args:
+            fragment_pairs: batch of pair of fragments.
+
+        Returns:
+            The preference probability for the first fragment for all fragment pairs
+            given by the network(s).
+            Shape: (num_fragment_pairs, ) for models with single reward networks and
+                   (num_fragment_pairs, num_networks) for ensemble of networks.
+
+        """
+        if self.is_ensemble:
+            probs = th.empty(
+                len(fragment_pairs),
+                self.model.num_members,
+                dtype=th.float32,
+            )
+        else:
+            probs = th.empty(len(fragment_pairs), dtype=th.float32)
+        for i, fragment in enumerate(fragment_pairs):
+            frag1, frag2 = fragment
+            trans1 = rollout.flatten_trajectories([frag1])
+            trans2 = rollout.flatten_trajectories([frag2])
+            rews1 = self.rewards(trans1)
+            rews2 = self.rewards(trans2)
+            probs[i] = self.probability(rews1, rews2)
+
+        return probs
+
+    def rewards(
+        self,
+        transitions: Transitions,
+    ) -> th.Tensor:
+        """Computes the reward for all transitions.
+
+        Args:
+            transitions: batch of obs-act-obs-done for a fragment of a trajectory.
+
+        Returns:
+            The reward given by the network(s) for all the transitions.
+            Shape: (num_transitions, ) for Single reward network and
+                   (num_transitions, num_networks) for ensemble of networks.
+        """
+        state = transitions.obs
+        action = transitions.acts
+        next_state = transitions.next_obs
+        done = transitions.dones
+        if self.is_ensemble:
```

Comment:
Why do we go via `predict_processed_all` for ensemble but `self.model` for no ensemble?

In particular I'm worried that `predict_processed_all` might apply normalization, which we do not want when computing probabilities (only for RL training).

@levmckinney I think you disabled forward on ensemble so might have some thoughts here on what the right thing to do is.

Reply:
Also, `predict_processed_all` returns NumPy arrays that have no gradients attached to them. This should not work for training right?

For now, I would recommend `rews = th.stack([member(state, action, next_state, done) for member in self.model.memeber], dim=-1)`. Second option we could bring back the forward_all method as a shorthand for this. Third option, we could make the reward `ensemble.forward` return a tensor of shape `(batch_size, num_members)` with all the rewards.